File: service/tests/data/world.rs
use crate::service::Seedable;
use chrono::{DateTime, Timelike, Utc};
use postgres::types::ToSql;
use uuid::Uuid;

/// A world that can be seeded into the database
#[derive(Debug, Clone)]
pub struct SeedWorld {
    /// The ID of the world
    pub world_id: Uuid,
    /// The version of the world
    pub version: Uuid,
    /// When the world was created
    pub created: DateTime<Utc>,
    /// When the world was last updated
    pub updated: DateTime<Utc>,

    /// The name of the world
    pub name: String,
    /// The description of the world
    pub description: String,
    /// The URL Slug of the world
    pub url_slug: String,
    /// The User ID of the owner of the world
    pub owner: Uuid,
}

impl Default for SeedWorld {
    fn default() -> Self {
        let now = Utc::now().with_nanosecond(0).unwrap();

        Self {
            world_id: Uuid::new_v4(),
            version: Uuid::new_v4(),
            created: now,
            updated: now,

            name: format!("name_{}", Uuid::new_v4()),
            description: format!("Description {}", Uuid::new_v4()),
            url_slug: format!("url_slug_{}", Uuid::new_v4()),
            owner: Uuid::new_v4(),
        }
    }
}

impl Seedable for SeedWorld {
    /// Generate the SQL needed to insert the seeded record into the database
    ///
    /// # Returns
    /// The SQL
    fn sql(&self) -> &str {
        "INSERT INTO worlds(world_id, version, created, updated, name, description, url_slug, owner_id)
        VALUES($1, $2, $3, $4, $5, $6, $7, $8)"
    }

    /// Generate the binds needed to insert the seeded record into the database
    ///
    /// # Returns
    /// The binds
    fn binds(&self) -> Vec<&(dyn ToSql + Sync)> {
        vec![
            &self.world_id,
            &self.version,
            &self.created,
            &self.updated,
            &self.name,
            &self.description,
            &self.url_slug,
            &self.owner,
        ]
    }
}

File: e2e/cypress/data/world.js
import { SeedData } from "./seed";
import { v4 as uuidv4 } from "uuid";

/**
 * Seed Data representing a world
 */
export class World extends SeedData {
  /**
   * Construct the world
   */
  constructor() {
    super();
    this._worldId = uuidv4();
    this._version = uuidv4();
    this._created = new Date();
    this._updated = new Date();
    this._name = uuidv4();
    this._description = "";
    this._urlSlug = uuidv4();
    this._ownerId = uuidv4();
  }

  /**
   * Set the name of this world
   * @param {string} value The new value for the name
   */
  withName(value) {
    this._name = value;
    return this;
  }

  /**
   * Set the description of this world
   * @param {string} value The new value for the description
   */
  withDescription(value) {
    this._description = value;
    return this;
  }

  /**
   * Set the URL Slug of this world
   * @param {string} value The new value for the URL Slug
   */
  withUrlSlug(value) {
    this._urlSlug = value;
    return this;
  }

  /**
   * Set the Owner ID of this world
   * @param {string} value The new value for the Owner ID
   */
  withOwnerId(value) {
    this._ownerId = value;
    return this;
  }

  async sql() {
    return "INSERT INTO worlds(world_id, version, created, updated, name, description, url_slug, owner_id) VALUES($1, $2, $3, $4, $5, $6, $7, $8)";
  }

  async binds() {
    return [
      this._worldId,
      this._version,
      this._created,
      this._updated,
      this._name,
      this._description,
      this._urlSlug,
      this._ownerId,
    ];
  }
}

async function hashPassword(password) {
  const { encoded } = await argon2.hash({
    pass: password,
    salt: "saltsaltsaltsaltsaltsaltsaltsalt",
  });
  return encoded;
}
